### ISSUE TITLE
Is it possible to make ExoPlayer playback transitions flow backwards?

### ISSUE THREAD

**Issue by GambitDev** [google/ExoPlayer#9720]
I'm building a media app with ExoPlayer, and I want the flow of the queue to flow backwards
Meaning the first item on the list would be the last on the queue (this is a customer request)

I am working with MediaSessionConnector and have been successful in changing it's response to media action so that "next" button would play the previous track.
But when a track is over it is automatically transitions to the next one on the list and I can't seem to figure out how to change this.

The app also supports Android Auto and Wear OS so the change must be internal so it would reflect properly there (and also with the notification controller)

This is how I define the MediaSessionConnector:

```kotlin
private fun prepareMediaSessionConnector() {
        mediaSessionConnector = MediaSessionConnector(mediaSession)
        mediaSessionConnector.setCustomActionProviders(
            BackCustomAction(),
            ForwardCustomAction()
        )

        mediaSessionConnector.setPlaybackPreparer(getMusicPlaybackPreparer())
        mediaSessionConnector.setQueueNavigator(
            object : TimelineQueueNavigator(mediaSession) {
                override fun getMediaDescription(
                    player: Player,
                    windowIndex: Int
                ): MediaDescriptionCompat {
                    return playerDataViewModel.getMediaDescriptionByWindowIndex(
                        windowIndex,
                        lastPodcastToRequestChapters,
                        isPlayingRadio
                    )
                }
            }
        )

        mediaSessionConnector.setControlDispatcher (
            object : DefaultControlDispatcher() {
                override fun dispatchNext(player: Player): Boolean {
                    if (exoPlayer.hasPreviousWindow()) {
                        exoPlayer.seekToPrevious()
                    } else {
                        Toast.makeText(this@MusicService, "No more items", Toast.LENGTH_LONG).show()
                    }
                    return true
                }

                override fun dispatchPrevious(player: Player): Boolean {
                    if (exoPlayer.hasNextWindow()) {
                        exoPlayer.seekToNext()
                    } else {
                        Toast.makeText(this@MusicService, "No more items", Toast.LENGTH_LONG).show()
                    }
                    return true
                }
            }
        )

        mediaSessionConnector.setPlayer(exoPlayer)
    }
```

In short, I want to override the logic of the automatic play next action.

I will gladly share more code or information if this isn't enough.

Any help would be appreciated, thanks!

**Comment by tonihei** [comment]
You can achieve this by setting a custom shuffle order in ExoPlayer and turn shuffling on:

```
exoPlayer.setShuffleOrder(new CustomShuffleOrder());
exoPlayer.setShuffleModeEnabled(true);
```

`CustomShuffleOrder` could be  implemented exactly as our test [`FakeShuffleOrder`](https://github.com/google/ExoPlayer/blob/release-v2/testutils/src/main/java/com/google/android/exoplayer2/testutil/FakeShuffleOrder.java) which plays everything in opposite order. If you need more fine-grained influence about the playback order, you could also use `DefaultShuffleOrder` and provide any ordered list of indices you would like to use.

**Comment by GambitDev** [comment]
@tonihei 
This is a great!
Thanks!

**Comment by tonihei** [comment]
Closing because the question was answered.
